File: acmOtherMaterials/Ulm Local Contest/2003/solution/assist.pas
(* Problem   Assistance Required
** Algorithm Precalculation
** Runtime   O(n*l[n])
** Author    Walter Guttmann
** Date      16.07.2000
*)

program assist;

const maxl: Longint = 33810;

var f: Text;
    l: array [0..4095] of Longint;
    b: array [0..33810] of Boolean;
    i, m, n, start: Longint;

begin
  (* precalculate all lucky numbers (the 3000th one is 33809) by simulation
  ** this would also suit the "Freiburg Method" for larger values of n
  *)
  for i := 0 to maxl do
    b[i] := True;
  start := 1;
  for n := 1 to 3000 do begin
    start := start+1;
    while not b[start] do
      start := start+1;
    l[n] := start;
    m := start;
    while m < maxl do begin
      b[m] := False;
      for i := 1 to start do begin
        m := m+1;
        while (m < maxl) and not b[m] do
          m := m+1
      end
    end
  end;
  Assign(f, 'assist.in');
  Reset(f);
  Read(f, n);
  while n <> 0 do begin
    WriteLn(l[n]);
    Read(f, n)
  end;
  Close(f)
end.

File: acmOtherMaterials/Ulm Local Contest/2003/solution/assist.java
// Problem   Assistance Required
// Algorithm Precalculation
// Runtime   O(n*l[n])
// Author    Walter Guttmann
// Date      16.07.2000

import java.io.*;

public class assist
{
  public static void main(String[] arg) throws Exception
  {
    final int maxl = 33810;
    int[] l = new int[4096];
    boolean b[] = new boolean[maxl];
    // precalculate all lucky numbers (the 3000th one is 33809) by simulation
    // this would also suit the "Freiburg Method" for larger values of n
    for (int i=0 ; i<maxl ; i++)
      b[i] = true;
    for (int n=1,start=1 ; n<=3000; n++)
    {
      ++start;
      while (!b[start]) ++start;
      l[n] = start;
      for (int m=start ; m<maxl ; )
      {
        b[m] = false;
        for (int i=0 ; i<start ; i++)
        {
          ++m;
          while (m<maxl && !b[m]) ++m;
        }
      }
    }
    StreamTokenizer st = new StreamTokenizer (new FileReader ("assist.in"));
    while (true)
    {
      st.nextToken();
      int n = (int) st.nval;
      if (n == 0) break;
      System.out.println(l[n]);
    }
  }
}

